File: 3871-count-commas-in-range-ii/3871-count-commas-in-range-ii.js
/**
 * @param {number} n
 * @return {number}
 */
var countCommas = function(n) {
    let total = 0;
    let start = 1000;
    let commas = 1;

    while (start <= n) {
        let end = Math.min(n, start * 1000 - 1);

        let count = end - start + 1;
        total += count * commas;

        // JavaScript handles safe integers up to 2^53 - 1. 
        // If n can exceed 9 quadrillion, use BigInt instead.
        start *= 1000;
        commas++;
    }

    return total;
};

File: 3871-count-commas-in-range-ii/3871-count-commas-in-range-ii.java
class Solution {
    public long countCommas(long n) {
        long total = 0;
        long start = 1000;
        long commas = 1;

        while (start <= n) {
            long end = Math.min(n, start * 1000 - 1);

            long count = end - start + 1;
            total += count * commas;

            start *= 1000;
            commas++;
        }

        return total;
    }
}